Update bind service directive to use common components/service. Bump Versions

jeff-phillips-18 opened the pull request:
Bumps version of origin-web-common to v0.0.26
Bumps version of origin-web-catalog to v0.0.15

Uses common pages for binding in the bindService directive
Updates the process template dialog to use the angular patternfly wizard (keeps consistent styling)

Fixes #1465

spadgett commented on app/scripts/controllers/landingPage.js:
```diff
@@ -51,10 +52,21 @@ angular.module('openshiftConsole')
+          var alert = {
+            name: 'error-loading-catalog-items',
+            data: {
+              type: 'warning',
```
Suggest an error here rather than a warning.

I'm surprised DataService doesn't show an error if the list fails. Won't there be two errors?

jeff-phillips-18 replied:
Will update to an error. the Catalog service getCatalogItems does not add any alerts/notifications as it does not set the errorNotification option in its calls to DataService.list

spadgett replied:
Hm, `errorNotification` should default to true.

https://github.com/openshift/origin-web-common/blob/master/src/services/dataService.js#L950

jeff-phillips-18 replied:
Ah, OK. I will remove the addNotification call.

jeff-phillips-18 replied:
Done.

spadgett commented on app/scripts/directives/bindService.js:
```diff
@@ -70,28 +73,46 @@ function BindService($filter,
+
+  ctrl.$onInit = function() {
+    ctrl.serviceSelection = {};
+    var formStepId = (ctrl.target.kind === 'Instance') ? 'applications' : 'services';
```
Would be it be simpler to have one step ID and then just change the step indicator label? I don't see where the ID is checked anywhere.

jeff-phillips-18 replied:
Done.

spadgett commented on app/scripts/directives/bindService.js:
```diff
@@ -151,89 +177,33 @@ function BindService($filter,
-        group: 'servicecatalog.k8s.io',
-        resource: 'bindings'
-      }, _.get(ctrl.binding, 'metadata.name'), context, function(binding) {
+      DataService.watchObject(BindingService.bindingResource, _.get(ctrl.binding, 'metadata.name'), context, function(binding) {
```
Not a new problem, but don't we need to be removing the watch when on scope destroy?

jeff-phillips-18 replied:
Done.

spadgett commented on app/scripts/filters/resources.js:
```diff
@@ -1351,8 +1316,21 @@ angular.module('openshiftConsole')
     };
+  })
+  // TODO: these two filters are defined in origin-web-catalog but are not available, need to figure out why or move to common
+  .filter('applicationHasDeployment', function ($filter) {
```
We have both filters already in origin-web-console

https://github.com/openshift/origin-web-console/blob/master/app/scripts/filters/resources.js#L719-L728

Although if we move them to common it might be better to remove the dependency on DeploymentsService so we don't have to move the entire service.

jeff-phillips-18 replied:
That dependency is why I didn't move them to common but added similar to catalog without the dependency. I was hesitant to add to common removing the dependency at this time.

spadgett replied:
OK I think we should look at it as a follow on to avoid having the same filters in multiple places.

jeff-phillips-18 replied:
Added issue https://github.com/openshift/origin-web-common/issues/82

spadgett commented on app/views/directives/process-template-dialog.html:
```diff
@@ -1,74 +1,49 @@
+         on-finish="$ctrl.close()"
+         on-cancel="$ctrl.close()"
+         wizard-done="$ctrl.wizardDone">
+      <div pf-wizard-step ng-repeat="step in $ctrl.steps track by $index"
```
`track by step.id` might be better?

jeff-phillips-18 replied:
Done.

spadgett commented:
Template dialog is empty the second time it's opened:

![openshift web console 2017-05-23 17-39-56](https://cloud.githubusercontent.com/assets/1167259/26377595/ecac0e74-3fde-11e7-9658-2f51e6d0f16d.png)

spadgett commented:
(Only after instantiating the template.)

spadgett commented:
I see this runtime error after canceling the template dialog. Appears to be a new problem.

```
angular.js:14199 TypeError: Cannot read property 'metadata' of null
    at Object.createFromTemplateURL (navigate.js:106)
    at resources.js:491
    at fn (eval at compile (angular.js:15126), <anonymous>:4:471)
    at regularInterceptedExpression (angular.js:16354)
    at expressionInputsWatch (angular.js:16282)
    at Scope.$digest (angular.js:17828)
    at Scope.$apply (angular.js:18102)
    at HTMLButtonElement.<anonymous> (angular-patternfly.js:9090)
    at HTMLButtonElement.dispatch (jquery.js:4435)
    at HTMLButtonElement.elemData.handle (jquery.js:4121)
```

spadgett commented:
Not related to this change, but I also noticed https://github.com/openshift/origin-web-catalog/issues/245 when testing.

spadgett commented:
The vertical divider extends too far down now (cc @rhamilto)

![openshift web console 2017-05-23 17-56-49](https://cloud.githubusercontent.com/assets/1167259/26378160/49c3e918-3fe1-11e7-884f-ba12cdc7c160.png)

spadgett commented:
We're showing the back button always on wizard now. Previously we were hiding it for two-step wizard because it was always disabled.

spadgett commented:
Seeing some other layout problems as well :/

![openshift web console 2017-05-23 18-01-16](https://cloud.githubusercontent.com/assets/1167259/26378298/f015a68a-3fe1-11e7-9453-c017512ca416.png)

spadgett commented:
Disregard that last screenshot (which I've now deleted). I accidentally switched branches.

spadgett commented:
I have a fix for the kebab not working:

https://github.com/openshift/origin-web-common/pull/81

jeff-phillips-18 commented:
@spadgett Fixed the issues you reported. Modified the structure a bit so the wizard would get re-created when re-displayed (resolves the blank page issue). Added a check for template in the process-template.html file since it still exists momentarily and updated when the template is set to null on close.

benjaminapetersen commented on app/scripts/directives/bindService.js:
```diff
@@ -70,28 +74,45 @@ function BindService($filter,
+  };
 
+  var showResults = function() {
+    if (validityWatcher) {
```
Mind explaining the validityWatcher here?  I'm curious.  I don't know that I've seen a watch stored, explicitly called, and nulled out like this before.

jeff-phillips-18 replied:
The validity watcher is added to know when the form is valid, and we then update the next button enablement. It must be removed when we leave the page so we keep a handle to it and call it when the  page changes or the component is destroyed.

benjaminapetersen commented on app/scripts/directives/bindService.js:
```diff
@@ -151,89 +177,36 @@ function BindService($filter,
-  ctrl.groupByKind = function(object) {
-    return humanizeKind(object.kind);
   };
 
```
nit: I tend to prefer `$onInit` at the beginning and `$onDestroy` at the end to make it easy/consistent to find the setup/tear down code. Not a big issue, just a pattern preference.

jeff-phillips-18 replied:
I like to keep the ctrl.$... functions together

benjaminapetersen commented on app/styles/_catalog.less:
```diff
@@ -1,8 +1,37 @@
+    overflow-y: auto;
+    padding-right: 0;
+    .scroll-shadows-vertical(65%, 0.25);
+    _::-webkit-full-page-media, _:future, :root & { // only target Safari
```
gnarly selector :smile:

rhamilto commented on app/styles/_catalog.less:
```diff
@@ -1,8 +1,37 @@
 @landing-search-area-height: 52px;
+@bind-service-page-height: 310px;
```
Nit:  alphabetization

rhamilto commented on app/styles/_catalog.less:
```diff
@@ -1,8 +1,37 @@
 
-.bind-service-wizard .wizard-pf-contents {
-  display: block;
+.wizard-pf-footer.wizard-pf-position-override {
```
Nit: alphabetization here and the rest of the rules you are adding

spadgett commented on app/scripts/controllers/landingPage.js:
```diff
@@ -51,10 +52,20 @@ angular.module('openshiftConsole')
+              message: errorMessage
+            }
+          };
+          AlertMessageService.addAlert(alert);
```
Adding the alert to `AlertMessageService` will show the message on the next page the user navigates to, which I don't think is what we want.

I suggest we change the `Catalog` service to pass `{ errorNotification: false }` and then show its formatted message here. In the meantime, we could either not do anything with the message or possibly just show separate alerts until we make the catalog change?

jeff-phillips-18 replied:
Probably best to have duplicates rather than none at all. I will add the notification back here, remove the alert here and file an issue for catalog to not add notifications.

spadgett replied:
Thanks, works for me

jeff-phillips-18 commented:
@rhamilto Fixed the alphabetization issues, thanks.

spadgett commented:
The extra space between the buttons here looks odd to me when the back button is hidden.

![openshift web console 2017-05-24 12-24-25](https://cloud.githubusercontent.com/assets/1167259/26414087/0c8f4088-407c-11e7-905b-14030c311c3c.png)

spadgett commented:
Also it looks like the Bind button text is not correctly centered?

jeff-phillips-18 commented:
Fixed the spacing for the buttons, added an override for not until angular patternfly fix can be made (https://github.com/patternfly/angular-patternfly/issues/469)

The button text is also an angular patternfly issue adding an extra space after the title. This we will need to wait for a fix on (https://github.com/patternfly/angular-patternfly/issues/468)

spadgett commented:
[merge]
